libs/blocks/gnav/gnav-contextual-search.js

```diff
@@ -0,0 +1,88 @@
+import { fetchBlogArticleIndex } from '../article-feed/article-feed.js';
+import { getArticleTaxonomy, buildArticleCard } from '../article-feed/article-helpers.js';
+import { createTag } from '../../utils/utils.js';
+
+function highlightTextElements(terms, elements) {
+  elements.forEach((element) => {
+    const matches = [];
+    const { textContent } = element;
+    const lowerCaseTextContent = textContent.toLowerCase();
+    terms.forEach((term) => {
+      const offset = lowerCaseTextContent.indexOf(term.toLowerCase());
+      if (offset >= 0) {
+        matches.push({ offset, term });
+      }
+    });
+    matches.sort((a, b) => a.offset - b.offset);
+    let currentIndex = 0;
+    const fragment = matches.reduce((acc, { offset, term }) => {
+      const textBefore = textContent.substring(currentIndex, offset);
+      if (textBefore) {
+        acc.appendChild(document.createTextNode(textBefore));
+      }
+      const markedTerm = createTag('mark', { class: 'gnav-search-highlight' }, term);
+      acc.appendChild(markedTerm);
+      currentIndex = offset + term.length;
+      return acc;
+    }, document.createDocumentFragment());
+    const textAfter = textContent.substring(currentIndex);
+    if (textAfter) {
+      fragment.appendChild(document.createTextNode(textAfter));
+    }
+    element.innerHTML = '';
+    element.appendChild(fragment);
+  });
+}
+
+export default async function onSearchInput({ value, resultsEl, searchInputEl, advancedSearchEl }) {
+  if (!value.length) {
+    resultsEl.innerHTML = '';
+    searchInputEl.classList.remove('gnav-search-input--isPopulated');
+    return;
+  }
+
+  resultsEl.classList.remove('no-results');
+  searchInputEl.classList.add('gnav-search-input--isPopulated');
+
+  const limit = 12;
+  const terms = value.toLowerCase().split(' ').filter(Boolean);
+  if (!terms.length) return;
+
+  const { data: articles } = await fetchBlogArticleIndex();
```

Comment:
So contextual search is specific to blog articles?  Will we want to expand that for other types of content?

Reply:
I will let @auniverseaway answer this. But, I think we should keep the scope of this PR for blog articles.

Reply:
ABP Documentation would like to use this contextual search feature, so it'd definitely be great if this could be applied to other types of content. The blog index at https://main--blog--adobecom.hlx.page/query-index.json?limit=500&offset=2000 doesn't look like it includes any custom blog article specific things, so I'm assuming it should be possible?

Reply:
What's the decision here? 

@saugatmalla What would need to be done for ABP or other sites to use this feature?

@dstrong23 This is the Contextual Search feature I mentioned. 

cc: @chrischrischris

Reply:
@rgclayton This is searching within the query-index of each project's contentRoot. I believe, we don't need to change any logic, except the namings that have blog (which may have been confused as being specific to blog). 

@honstar @chrischrischris LMK if I am wrong.

Reply:
I agree with @saugatmalla : afaict, the proposed changes in this PR merely use a project's `query-index` as a base to perform the contextual search on the client-side.

cc: @chrischrischris @auniverseaway @rgclayton